### ISSUE TITLE
Strange Tensorflow metrics behavior

### ISSUE THREAD

**Issue by AlexZhurkevich** [NVIDIA/DALI#2631]
Hello everybody,
Having the same model, same tfrecords but different input pipelines based on tensorflow vs DALI produces strange metrics behavior. I wouldn't be surprised if preprocessing was different but both of them do simple greyscale read, float conversion and normalization.

Let me start by showing common pieces of the program (GPU mem limiting, IoU, Model), just in case it will clear a picture:
```
BATCH_SIZE=12
gpus = tf.config.experimental.list_physical_devices('GPU')
if gpus:
    # Restrict TensorFlow to only allocate 5GB of memory on the first GPU
    try:
        tf.config.experimental.set_virtual_device_configuration(
            gpus[0],
            [tf.config.experimental.VirtualDeviceConfiguration(memory_limit=5120)])
        logical_gpus = tf.config.experimental.list_logical_devices('GPU')
        print(len(gpus), "Physical GPUs,", len(logical_gpus), "Logical GPUs")
    except RuntimeError as e:
        # Virtual devices must be set before GPUs have been initialized
        print(e)

#IoU function
def iou(y_true, y_pred):
    def f(y_true, y_pred):
        #print(y_pred)
        intersection = (y_true * y_pred).sum()
        union = y_true.sum() + y_pred.sum() - intersection
        x = (intersection + 1e-15) / (union + 1e-15)
        x = x.astype(np.float32)
        return x
    return tf.numpy_function(f, [y_true, y_pred], tf.float32)

#Model
def get_model(size, kernel_size):
    inputs = keras.Input((size, size, 1))

    ### [First half of the network: downsampling inputs] ###

    # Entry block
    x = layers.Conv2D(16, (kernel_size, kernel_size), strides=2, padding="same")(inputs)
    x = layers.BatchNormalization()(x)
    x = layers.Activation("relu")(x)

    previous_block_activation = x  # Set aside residual

    # Blocks 1, 2, 3 are identical apart from the feature depth.
    for filters in [32, 64, 128, 256]:
        x = layers.Activation("relu")(x)
        x = layers.SeparableConv2D(filters, (kernel_size, kernel_size), padding="same")(x)
        x = layers.BatchNormalization()(x)

        x = layers.Activation("relu")(x)
        x = layers.SeparableConv2D(filters, (kernel_size, kernel_size), padding="same")(x)
        x = layers.BatchNormalization()(x)

        x = layers.MaxPooling2D((kernel_size, kernel_size), strides=2, padding="same")(x)

        # Project residual
        residual = layers.Conv2D(filters, 1, strides=2, padding="same")(
            previous_block_activation
        )
        x = layers.add([x, residual])  # Add back residual
        previous_block_activation = x  # Set aside next residual

    ### [Second half of the network: upsampling inputs] ###

    for filters in [256, 128, 64, 32, 16]:
        x = layers.Activation("relu")(x)
        x = layers.Conv2DTranspose(filters, (kernel_size, kernel_size), padding="same")(x)
        x = layers.BatchNormalization()(x)

        x = layers.Activation("relu")(x)
        x = layers.Conv2DTranspose(filters, (kernel_size, kernel_size), padding="same")(x)
        x = layers.BatchNormalization()(x)

        x = layers.UpSampling2D(2)(x)

        # Project residual
        residual = layers.UpSampling2D(2)(previous_block_activation)
        residual = layers.Conv2D(filters, 1, padding="same")(residual)
        x = layers.add([x, residual])  # Add back residual
        previous_block_activation = x  # Set aside next residual

    # Add a per-pixel classification layer
    x = layers.Conv2D(1, 3, padding="same")(x)
    outputs = layers.Activation("sigmoid", dtype='float32', name='predictions')(x)

    # Define the model
    model = keras.Model(inputs, outputs)
    return model

metrics = ["acc", iou]
model = get_model(512, 3)         
model.compile(optimizer="adam", loss="binary_crossentropy", metrics=metrics)
model.summary()
model.fit(your_pipeline,
          epochs=500,
          steps_per_epoch=359/BATCH_SIZE)
```
Let's get to differences, first I will show tested tensorflow 2 preprocessing pipeline:
```
def read_tfrecord(record):
    keys_to_features = {
        'image/height': tf.io.FixedLenFeature([], tf.int64),
        'image/width': tf.io.FixedLenFeature([], tf.int64),
        'image/channels': tf.io.FixedLenFeature([], tf.int64),
        'image/class/mask': tf.io.FixedLenFeature([], tf.string),
        'image/image_filename': tf.io.FixedLenFeature([], tf.string),
        'image/mask_filename': tf.io.FixedLenFeature([], tf.string),
        'image/encoded': tf.io.FixedLenFeature([], tf.string)
    }
    sample =  tf.io.parse_single_example(record, keys_to_features)
    image = tf.io.decode_jpeg(sample['image/encoded'], channels=1)
    mask = tf.io.decode_jpeg(sample['image/class/mask'], channels=1)
    image = tf.cast(image, tf.float32) / 255
    mask = tf.cast(mask, tf.float32) / 255
    return image, mask

def get_batched_train_dataset(BATCH_SIZE, filenames):
    files = tf.data.Dataset.list_files(filenames)
    dataset = files.interleave(tf.data.TFRecordDataset, num_parallel_calls=tf.data.AUTOTUNE)
    dataset = dataset.map(read_tfrecord, num_parallel_calls=tf.data.AUTOTUNE)
    dataset = dataset.batch(BATCH_SIZE, drop_remainder=False)
    dataset = dataset.prefetch(tf.data.AUTOTUNE)
    dataset = dataset.repeat()
    return dataset
def get_training_dataset():
  return get_batched_train_dataset(BATCH_SIZE, "/media/*.tfrecord")

#Link it with model like so
model.fit(get_training_dataset(), .......)
```
Tensorflow 2 preprocessing pipeline output:
```
Epoch 1/500
2021-01-22 17:50:09.880671: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.8
2021-01-22 17:50:10.562331: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2021-01-22 17:50:10.744429: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.11
29/29 [==============================] - 22s 588ms/step - loss: 1.4820 - acc: 0.8756 - iou: 0.8893
Epoch 2/500
29/29 [==============================] - 16s 523ms/step - loss: 0.1093 - acc: 0.9622 - iou: 0.9591
Epoch 3/500
29/29 [==============================] - 16s 521ms/step - loss: 0.0936 - acc: 0.9627 - iou: 0.9577
Epoch 4/500
29/29 [==============================] - 16s 522ms/step - loss: 0.0870 - acc: 0.9656 - iou: 0.9639
Epoch 5/500
29/29 [==============================] - 14s 482ms/step - loss: 0.0945 - acc: 0.9642 - iou: 0.9615
Epoch 6/500
29/29 [==============================] - 15s 487ms/step - loss: 0.0950 - acc: 0.9631 - iou: 0.9624
Epoch 7/500
29/29 [==============================] - 15s 491ms/step - loss: 0.0745 - acc: 0.9666 - iou: 0.9656
Epoch 8/500
29/29 [==============================] - 15s 490ms/step - loss: 0.0881 - acc: 0.9617 - iou: 0.9562
Epoch 9/500
29/29 [==============================] - 14s 481ms/step - loss: 0.0761 - acc: 0.9641 - iou: 0.9598
Epoch 10/500
29/29 [==============================] - 14s 483ms/step - loss: 0.0811 - acc: 0.9631 - iou: 0.9594

```
Now we will run pipeline based on DALI code, it looks like this:
```
def natural_sort(sort_list):
    convert = lambda text: int(text) if text.isdigit() else text
    alphanum_key = lambda key: [ convert(c) for c in re.split('([0-9]+)', key) ]
    sort_list.sort(key=alphanum_key)
    return sort_list

#Sort your tfrecords along with their indexes, to correctly match the two
tfrecords = natural_sort(glob('/media/*.tfrecord'))
indexes = natural_sort(glob('/media/idx_files/*.idx'))

#Pipeline itself
class SegmantationPipeline(Pipeline):

    def __init__(self, batch_size, device, device_id=0, num_threads=12, exec_pipelined=False):
        super(SegmantationPipeline, self).__init__(
            batch_size, num_threads, device_id)
        
        #CPU or GPU
        self.device = device

        self.reader = ops.TFRecordReader(
            #List of paths to TFRecord files.
            path=tfrecords,
            #List of paths to index files. There should be one index file for every TFRecord file.
            index_path=indexes,
            #A dictionary that maps names of the TFRecord features to extract to the feature type.
            features={
                'image/height': tfrec.FixedLenFeature([], tfrec.int64, 0),
                'image/width': tfrec.FixedLenFeature([], tfrec.int64, 0),
                'image/channels': tfrec.FixedLenFeature([], tfrec.int64, 0),
                'image/class/mask': tfrec.FixedLenFeature([], tfrec.string, ""),
                'image/image_filename': tfrec.FixedLenFeature([], tfrec.string, ""),
                'image/mask_filename': tfrec.FixedLenFeature([], tfrec.string, ""),
                'image/encoded': tfrec.FixedLenFeature([], tfrec.string, "")
            },
            #Determines whether to randomly shuffle data.
            random_shuffle=True,
            #Size of the buffer that is used for shuffling.
            initial_fill=2048,
            #Specifies the number of batches to be prefetched by the internal Loader.
            prefetch_queue_depth=5,
            #Determines whether the accessed data should be read ahead.
            read_ahead=True
        )

        #Decode image
        self.decode = ops.ImageDecoder(
            device='mixed' if device == 'gpu' else 'cpu',
            output_type=types.GRAY)
        
        #Normalize image
        self.norm = ops.Normalize(
            device=device,
            dtype=types.FLOAT)

    def define_graph(self):
        inputs = self.reader(name="Reader")
        images = self.decode(inputs["image/encoded"])
        masks = self.decode(inputs['image/class/mask'])

        if self.device == 'gpu':
            images = images.gpu()
            masks = masks.gpu()

        images = self.norm(images)
        masks = self.norm(masks)
        return (images, masks)
        
segmentation_pipeline = SegmantationPipeline(BATCH_SIZE, device='gpu', device_id=0)

shapes = (
    (BATCH_SIZE, 512, 512, 1),
    (BATCH_SIZE, 512, 512, 1))
dtypes = (
    tf.float32,
    tf.float32)

seg_set = dali_tf.DALIDataset(
    pipeline=segmentation_pipeline,
    batch_size=BATCH_SIZE,
    output_shapes=shapes,
    output_dtypes=dtypes,
    num_threads=12,
    device_id=0)

#Link it with model like so
model.fit(seg_set, .............)

```
DALI preprocessing pipeline output:
```
Epoch 1/500
2021-01-22 17:53:55.406816: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:116] None of the MLIR optimization passes are enabled (registered 2)
2021-01-22 17:53:55.435639: I tensorflow/core/platform/profile_utils/cpu_utils.cc:112] CPU Frequency: 2199995000 Hz
2021-01-22 17:53:55.970574: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcudnn.so.8
2021-01-22 17:53:56.682382: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublas.so.11
2021-01-22 17:53:56.854801: I tensorflow/stream_executor/platform/default/dso_loader.cc:49] Successfully opened dynamic library libcublasLt.so.11
29/29 [==============================] - 19s 475ms/step - loss: -344.2761 - accuracy: 0.0000e+00 - iou: 0.1339
Epoch 2/500
29/29 [==============================] - 14s 470ms/step - loss: -350035.1895 - accuracy: 0.0000e+00 - iou: 0.1841
Epoch 3/500
29/29 [==============================] - 14s 474ms/step - loss: -68822095.6613 - accuracy: 0.0000e+00 - iou: 0.1594
Epoch 4/500
29/29 [==============================] - 14s 471ms/step - loss: -4988786252.3871 - accuracy: 0.0000e+00 - iou: 0.1786
Epoch 5/500
29/29 [==============================] - 14s 469ms/step - loss: -154530704350.9677 - accuracy: 0.0000e+00 - iou: 0.1873
Epoch 6/500
29/29 [==============================] - 14s 478ms/step - loss: -2743911702263.7417 - accuracy: 0.0000e+00 - iou: 0.1525
Epoch 7/500
29/29 [==============================] - 14s 464ms/step - loss: -29622208747189.6758 - accuracy: 0.0000e+00 - iou: 0.1663
Epoch 8/500
29/29 [==============================] - 14s 464ms/step - loss: -220701027382701.4062 - accuracy: 0.0000e+00 - iou: 0.1548
Epoch 9/500
29/29 [==============================] - 14s 469ms/step - loss: -1385831652316853.7500 - accuracy: 0.0000e+00 - iou: 0.1716
Epoch 10/500
29/29 [==============================] - 14s 466ms/step - loss: -7191044754049090.0000 - accuracy: 0.0000e+00 - iou: 0.1509
```
You can clearly see that something wrong happens with metrics calculation. At first I though, maybe there is a mismatch in DALI between images and masks, I have built the pipeline separately, saved masks and images, they matched. 
Later I figured that in tensorflow 2 pipeline, when you `print(dataset)` the output is `<RepeatDataset shapes: ((None, None, None, 1), (None, None, None, 1)), types: (tf.float32, tf.float32)>`, so I've changed shapes and dtypes accordingly:
```
shapes = (
    (None, None, None, 1),
    (None, None, None, 1))
dtypes = (
    tf.float32,
    tf.float32)
```
but that also didnt change the behavior. I also let it run for some time, but that didnt help:
```
Epoch 94/500
29/29 [==============================] - 14s 472ms/step - loss: -169150229628956000.0000 - accuracy: 0.0000e+00 - iou: -2.4261e-14
Epoch 95/500
29/29 [==============================] - 14s 472ms/step - loss: -126504624957440528.0000 - accuracy: 0.0000e+00 - iou: 4.7781e-15
Epoch 96/500
29/29 [==============================] - 14s 473ms/step - loss: -127626403825844224.0000 - accuracy: 0.0000e+00 - iou: 2.0609e-16
Epoch 97/500
29/29 [==============================] - 14s 471ms/step - loss: -60760280492789360.0000 - accuracy: 0.0000e+00 - iou: 0.0229
Epoch 98/500
29/29 [==============================] - 14s 478ms/step - loss: -37574441360653280.0000 - accuracy: 0.0000e+00 - iou: 0.0387
Epoch 99/500
29/29 [==============================] - 14s 471ms/step - loss: -36856786512045880.0000 - accuracy: 0.0000e+00 - iou: 2.3980e-15
Epoch 100/500
29/29 [==============================] - 14s 469ms/step - loss: -66305316386470880.0000 - accuracy: 0.0000e+00 - iou: -2.9402e-15

```
Has anybody had this sort of an issue? Seems very unusual especially considering model and metrics are not changed, any help/recommendations are welcomed. 
Thanks you.

**Comment by JanuszL** [comment]
Hi,
I would dump data as NumPy arrays and compare them side by side.
The first thing that comes to my mind is that you return a value 0-255. from DALI pipeline. You should use CropMirrorNormalize and 255 as std, or just:
```
        images = images / 255.
        masks =images / 255.
```
ops.Normalize normalizes using calculates mean and std per sample (in your configuration).

**Comment by JanuszL** [comment]
Please reopen if it still doesn't work for you.
